warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java

```diff
@@ -1491,47 +1493,76 @@ public static Double getNodeSelectivity(Set<String> fieldNames, ShardQueryConfig
-                                throw new RuntimeException("Failed to validate lineage: Tree included a child with a null parent.");
-                            else
-                                log.error("Failed to validate lineage: Tree included a child with a null parent.");
-                            
-                            result = false;
+                            String message = "Tree included child " + child + " with a null parent";
+                            handleInvalidation(message, failHard, validation);
                         } else if (child.jjtGetParent() != node) {
-                            if (failHard)
-                                throw new RuntimeException("Failed to validate lineage:  Included a child with a conflicting parent.");
-                            else
-                                log.error("Failed to validate lineage:  Included a child with a conflicting parent.");
-                            
-                            result = false;
+                            String message = "Included a child " + child + " with conflicting parent. Expected " + node + " but was " + child.jjtGetParent();
+                            handleInvalidation(message, failHard, validation);
                         }
                         workingStack.push(child);
                     } else {
-                        if (failHard)
-                            throw new RuntimeException("Failed to validate lineage: Included a null child.");
-                        else
-                            log.error("Failed to validate lineage: Included a null child.");
-                        
-                        result = false;
+                        String message = "Included a null child under parent " + node;
+                        handleInvalidation(message, failHard, validation);
                     }
                 }
             }
         }
-        return result;
+        
+        return validation;
+    }
+    
+    private static void handleInvalidation(String message, boolean failHard, LineageValidation validation) {
+        if (failHard) {
+            throw new RuntimeException("Failed to validate lineage: " + message);
+        } else {
+            log.error("Failed to validate lineage: " + message);
+            validation.addInvalidation(message);
+        }
+    }
+    
+    public static class LineageValidation {
+        private final List<String> invalidations = new ArrayList<>();
+        
+        public boolean isValid() {
+            return invalidations.isEmpty();
+        }
+        
+        public void addInvalidation(String message) {
+            invalidations.add(message);
+        }
+        
+        public String getFormattedInvalidations() {
+            if (isValid()) {
+                return null;
+            } else if (invalidations.size() == 1) {
+                return invalidations.get(0);
+            } else {
```

Comment:
It looks like this else case also handles a single element gracefully. The else/if above could be dropped.